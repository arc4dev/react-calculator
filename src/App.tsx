import { useReducer } from 'react';

import DigitButton from './DigitButton';
import OperationButton from './OperationButton';
import ActionButton from './ActionButton';

const initialState: AppState = {
  previousOperand: null,
  currentOperand: '0',
  operator: null,
};

const reducer = (state: AppState, { type, payload }: ActionTypes) => {
  const computeResult = (operator: string, num1: number, num2: number) => {
    let computation: number;

    switch (operator) {
      case '+':
        computation = num1 + num2;
        break;
      case '-':
        computation = num1 - num2;
        break;
      case 'x':
        computation = num1 * num2;
        break;
      case 'รท':
        computation = num1 / num2;
        break;
      default:
        throw new Error(`Unknown operator: ${operator}`);
    }

    return String(computation);
  };

  switch (type) {
    case 'ADD_NUMBER':
      if (state.currentOperand.includes('.') && payload === '.') return state;
      if (state.currentOperand === '0' && payload === '0') return state;
      if (state.currentOperand === '0')
        return {
          ...state,
          currentOperand: payload,
        };

      return {
        ...state,
        currentOperand: `${state.currentOperand}${payload}`,
      };

    case 'CHOOSE_OPERATOR':
      if (state.currentOperand === '0' && state.previousOperand === null)
        return state;

      if (state.operator && state.currentOperand !== '0') {
        const computation = computeResult(
          state.operator,
          parseFloat(state.previousOperand as string),
          parseFloat(state.currentOperand)
        );

        return {
          ...state,
          previousOperand: computation,
          currentOperand: '0',
          operator: payload,
        };
      }

      if (state.currentOperand === '0')
        return {
          ...state,
          operator: payload,
        };
      console.log('yey');

      return {
        ...state,
        operator: payload,
        previousOperand: state.currentOperand,
        currentOperand: '0',
      };

    case 'CLEAR':
      return initialState;

    case 'DELETE_NUMBER':
      if (state.currentOperand.length === 1)
        return { ...state, currentOperand: '0' };

      if (state.currentOperand === '0') return state;

      return {
        ...state,
        currentOperand: state.currentOperand.slice(0, -1),
      };

    case 'COMPUTE':
      if (!state.operator) return state;

      const computation = computeResult(
        state.operator,
        parseFloat(state.previousOperand as string),
        parseFloat(state.currentOperand)
      );

      return {
        ...state,
        previousOperand: null,
        currentOperand: computation,
        operator: null,
      };

    default:
      return state;
  }
};

function App() {
  const [{ previousOperand, currentOperand, operator }, dispatch] = useReducer(
    reducer,
    initialState
  );

  return (
    <>
      <h1 className="heading">
        <span>Calc</span>ulator
      </h1>

      <div className="calculator">
        <div className="output">
          <span className="previous-operand">
            {previousOperand}
            {operator}
          </span>
          <p className="current-operand">{currentOperand}</p>
        </div>

        <ActionButton dispatch={dispatch} type="CLEAR" className="span-two">
          AC
        </ActionButton>
        <ActionButton dispatch={dispatch} type="DELETE_NUMBER">
          DEL
        </ActionButton>
        <OperationButton dispatch={dispatch}>รท</OperationButton>
        <DigitButton dispatch={dispatch}>1</DigitButton>
        <DigitButton dispatch={dispatch}>2</DigitButton>
        <DigitButton dispatch={dispatch}>3</DigitButton>
        <OperationButton dispatch={dispatch}>x</OperationButton>
        <DigitButton dispatch={dispatch}>4</DigitButton>
        <DigitButton dispatch={dispatch}>5</DigitButton>
        <DigitButton dispatch={dispatch}>6</DigitButton>
        <OperationButton dispatch={dispatch}>+</OperationButton>
        <DigitButton dispatch={dispatch}>7</DigitButton>
        <DigitButton dispatch={dispatch}>8</DigitButton>
        <DigitButton dispatch={dispatch}>9</DigitButton>
        <OperationButton dispatch={dispatch}>-</OperationButton>
        <DigitButton dispatch={dispatch}>0</DigitButton>
        <DigitButton dispatch={dispatch}>.</DigitButton>
        <ActionButton
          dispatch={dispatch}
          type="COMPUTE"
          className="span-two operator">
          =
        </ActionButton>
      </div>

      <footer className="footer">
        <span>&copy; 2023 Arkadiusz Sroczyk.</span>
      </footer>
    </>
  );
}

export default App;
