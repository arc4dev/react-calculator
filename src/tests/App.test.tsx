import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import App from '../App';

const performOperation = (operations: string[]) => {
  operations.forEach((operation) => {
    userEvent.click(screen.getByRole('button', { name: operation }));
  });
};

describe('App rendering', () => {
  it('renders Calculator', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Calculator operations', () => {
  it('should add two number properly', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['1', '+', '5', '+', '9', '=']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('15');
  });

  it('should subtract two number properly', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['8', '-', '5', '-', '1', '=']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('2');
  });

  it('should multiply two number properly', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['3', 'x', '4', 'x', '2', '=']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('24');
  });

  it('should divide two number properly', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['9', '0', 'รท', '5', '=']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('18');
  });

  it('should clear the result', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['9', 'รท', '3', 'AC', '0', 'AC']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('0');
  });

  it('should delete the last digit', () => {
    // arrange
    render(<App />);

    // act
    act(() => {
      performOperation(['5', '4', '2', 'DEL', 'DEL']);
    });

    // assert
    const currentResult = screen.getByTestId('current-result');
    expect(currentResult).toHaveTextContent('5');
  });
});
