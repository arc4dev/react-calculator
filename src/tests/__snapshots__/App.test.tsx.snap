// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering renders Calculator 1`] = `
Array [
  <h1
    className="heading"
  >
    <span>
      Calc
    </span>
    ulator
  </h1>,
  <div
    className="calculator"
  >
    <div
      className="output"
    >
      <span
        className="previous-operand"
      />
      <p
        className="current-operand"
        data-testid="current-result"
      >
        0
      </p>
    </div>
    <button
      className="action span-two"
      onClick={[Function]}
    >
      AC
    </button>
    <button
      className="action undefined"
      onClick={[Function]}
    >
      DEL
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      ÷
    </button>
    <button
      onClick={[Function]}
    >
      1
    </button>
    <button
      onClick={[Function]}
    >
      2
    </button>
    <button
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      x
    </button>
    <button
      onClick={[Function]}
    >
      4
    </button>
    <button
      onClick={[Function]}
    >
      5
    </button>
    <button
      onClick={[Function]}
    >
      6
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      7
    </button>
    <button
      onClick={[Function]}
    >
      8
    </button>
    <button
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      -
    </button>
    <button
      onClick={[Function]}
    >
      0
    </button>
    <button
      onClick={[Function]}
    >
      .
    </button>
    <button
      className="action span-two operator"
      onClick={[Function]}
    >
      =
    </button>
  </div>,
  <footer
    className="footer"
  >
    <span>
      © 2023 Arkadiusz Sroczyk.
    </span>
  </footer>,
]
`;

exports[`App renders Calculator 1`] = `
Array [
  <h1
    className="heading"
  >
    <span>
      Calc
    </span>
    ulator
  </h1>,
  <div
    className="calculator"
  >
    <div
      className="output"
    >
      <span
        className="previous-operand"
      />
      <p
        className="current-operand"
        data-testid="current-result"
      >
        0
      </p>
    </div>
    <button
      className="action span-two"
      onClick={[Function]}
    >
      AC
    </button>
    <button
      className="action undefined"
      onClick={[Function]}
    >
      DEL
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      ÷
    </button>
    <button
      onClick={[Function]}
    >
      1
    </button>
    <button
      onClick={[Function]}
    >
      2
    </button>
    <button
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      x
    </button>
    <button
      onClick={[Function]}
    >
      4
    </button>
    <button
      onClick={[Function]}
    >
      5
    </button>
    <button
      onClick={[Function]}
    >
      6
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      7
    </button>
    <button
      onClick={[Function]}
    >
      8
    </button>
    <button
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="operator"
      onClick={[Function]}
    >
      -
    </button>
    <button
      onClick={[Function]}
    >
      0
    </button>
    <button
      onClick={[Function]}
    >
      .
    </button>
    <button
      className="action span-two operator"
      onClick={[Function]}
    >
      =
    </button>
  </div>,
  <footer
    className="footer"
  >
    <span>
      © 2023 Arkadiusz Sroczyk.
    </span>
  </footer>,
]
`;
